pkg/backend/metrics.go

```diff
@@ -0,0 +1,203 @@
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+const (
+	metricPackage = "backend.openshiftcluster"
+	metricInstall = metricPackage + ".installphase"
+	enabled       = "Enabled"
+	disabled      = "Disabled"
+	custom        = "Custom"
+	defaultSet    = "Default"
+	unknown       = "unknown"
+	empty         = "empty"
+)
+
+func (ocb *openShiftClusterBackend) emitMetrics(log *logrus.Entry, doc *api.OpenShiftClusterDocument, operationType, provisioningState api.ProvisioningState, backendErr error) {
+	dimensions := map[string]string{
+		"operationtype":     operationType.String(),
+		"provisioningstate": provisioningState.String(),
+		"resulttype":        string(ocb.getResultType(backendErr)),
+	}
+
+	ocb.gatherCorrelationID(doc, dimensions)
+	ocb.gatherMiscMetrics(doc, dimensions)
+	ocb.gatherAuthMetrics(doc, dimensions)
+	ocb.gatherNetworkMetrics(doc, dimensions)
+	ocb.gatherNodeMetrics(doc, dimensions)
+
+	ocb.logMetricDimensions(log, dimensions)
+	ocb.m.EmitGauge(metricInstall, 1, dimensions)
+}
+
+func (ocb *openShiftClusterBackend) getResultType(backendErr error) utillog.ResultType {
+	var resultType utillog.ResultType
+	err, ok := backendErr.(*api.CloudError)
+	if ok {
+		resultType = utillog.MapStatusCodeToResultType(err.StatusCode)
+	}
+	return resultType
+}
+
+func (ocb *openShiftClusterBackend) logMetricDimensions(log *logrus.Entry, dimensions map[string]string) {
+	for metric, value := range dimensions {
+		log.Info(fmt.Sprintf("%s: %s = %s", metricPackage, metric, value))
+	}
+}
+
+func (m *openShiftClusterBackend) gatherCorrelationID(doc *api.OpenShiftClusterDocument, dimensions map[string]string) {
+	if doc.CorrelationData != nil {
+		dimensions["correlationid"] = doc.CorrelationData.CorrelationID
+		dimensions["client_requestid"] = doc.CorrelationData.ClientRequestID
+		dimensions["requestid"] = doc.CorrelationData.RequestID
+		dimensions["async_operationsid"] = doc.AsyncOperationID
+	} else {
+		dimensions["correlationid"] = empty
+		dimensions["client_requestid"] = empty
+		dimensions["async_operationsid"] = empty
+	}
+}
+
+func (ocb *openShiftClusterBackend) gatherMiscMetrics(doc *api.OpenShiftClusterDocument, dimensions map[string]string) {
+	dimensions["subscriptionid"] = ocb.env.SubscriptionID()
+	dimensions["resourceid"] = doc.ResourceID
+	dimensions["location"] = doc.OpenShiftCluster.Location
+	dimensions["resourcegroup"] = doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID
+	dimensions["name"] = doc.OpenShiftCluster.Name
+	dimensions["ocpversion"] = doc.OpenShiftCluster.Properties.ClusterProfile.Version
+	dimensions["workercount"] = strconv.FormatInt(int64(len(doc.OpenShiftCluster.Properties.WorkerProfiles)), 10)
+	dimensions["rpversion"] = doc.OpenShiftCluster.Properties.ProvisionedBy
+
+	for flag, feature := range doc.OpenShiftCluster.Properties.OperatorFlags {
+		dimensions[fmt.Sprintf("operator-%s", flag)] = feature
+	}
```

Comment:
Do we need this? I don't think users have any ability to influence changes to the default operator flags we put down, so these values will be the same for all cluster installs.

Reply:
It maybe helpful for when we do adminupdates, which will also emit these metrics.  
This gets called for creating, adminupdate, update, and deleting.

Reply:
I missed that part, that invalidates a lot of my feedback so far. I'll go through and fixup suggestions.

Reply:
Thank you.